### DISCUSSION TITLE
Why Request->normalizeAndFilterClientIps() has array_reverse($clientIps)?

### DISCUSSION THREAD

**Discussion by 99hops** [symfony/symfony#47453] [Q&A] [answered]
When there are multiple x-forwarded-for IPs like "10.10.10.10, 20.20.20.20" in the end of Request->normalizeAndFilterClientIps()
there is array_reverse() which results in ["20.20.20.20, 10.10.10.10"] 
```
// Now the IP chain contains only untrusted proxies and the client IP
return $clientIps ? array_reverse($clientIps) : [$firstTrustedIp];
```

then there is 
```
public function getClientIp(): ?string
{
    $ipAddresses = $this->getClientIps();

    return $ipAddresses[0];
}
```

which results in getClientIp() = "20.20.20.20".

Here is what happens in my case: I use HAProxy to forward (option forwardfor "X-Forwarded-For") requests to Kong which forwards to some service and getClientIp() returns the address of the HAProxy server instead of the actual IP $ipAddresses[0] because of the array_reverse().

```
framework:
    trusted_proxies: '10.5.0.1'
    trusted_headers: [ 'x-forwarded-for' ]
```
What is the reason for this array_reverse() to exists and is this the expected behavior?

**Comment by alamirault** [comment]
Original commit (9 years ago 😲 ) https://github.com/symfony/symfony/commit/75db8eb8a7fce5098715cc17b1e570d49ceabb93

````
   * The most trusted IP address is first, and the less trusted one last.
   * The "real" client IP address is the last one, but this is also the
   * less trusted one.
   ````

So, I think is this the expected behavior

**Reply by 99hops** [reply]
I don't know what to say about this :smiley_cat: 
`The "real" client IP address is the last one, but this is also the less trusted one.`
If this is the case and the list is reversed by array_reverse() I seems that "the less trusted one" is taken.

In my case there is CLIENT -> HAProxy -> KONG_GATEWAY -> SERVICE (Symfony) which results in x-forwarded-for: [CLIENT_IP, HAPROXY_IP] before array_reverse().

What do you think @alamirault ?

**Reply by MatTheCat** [reply]
Be aware the method you mention (`normalizeAndFilterClientIps`) also append the client IP to the array: https://github.com/symfony/symfony/blob/36de7feaab115296720aa617e2d05638a5ce9441/src/Symfony/Component/HttpFoundation/Request.php#L2012

**Comment by stof** [comment, marked as answer]
If you want to have `getClientIp` returning the IP of the browser, you would need to configure both KONG_GATEWAY and HAProxy as trusted proxies. This way, CLIENT will be the IP being returned (it would be trusted as it is the IP from which the request reached a trusted server).

Blindly trusting `X-Forwarded-For` would be a security risk, as nothing prevents a malicious user to add a `X-Forwarded-For` header when sending the request (pretending it forwards it on behalf of an IP belonging to your company network for instance).

**Reply by 99hops** [reply]
Thanks @stof! Adding both gateways leaves only the CLIENT.
